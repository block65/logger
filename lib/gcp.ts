import pino from 'pino';
import { isPlainObject } from './utils.js';

type GcpSeverities = 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL';

export const logLevelGcpSeverityMap = new Map<
  pino.Level | string,
  GcpSeverities
>([
  ['trace', 'DEBUG'],
  ['debug', 'DEBUG'],
  ['info', 'INFO'],
  ['warn', 'WARNING'],
  ['error', 'ERROR'],
  ['fatal', 'CRITICAL'],
]);

export function formatGcpLogObject(details: object): object {
  if (isPlainObject(details)) {
    if ('stack' in details) {
      const { stack, ...rest } = details;
      if (stack) {
        return {
          stack_trace: stack,
          ...rest,
        };
      }
    }

    if (
      ('err' in details && isPlainObject(details.err)) ||
      details.err instanceof Error
    ) {
      const { err, ...rest } = details;
      const { stack, ...restErr } = err;
      if (stack) {
        return {
          stack_trace: stack,
          ...rest,
          ...restErr,
        };
      }
    }
  }
  return details;
}
