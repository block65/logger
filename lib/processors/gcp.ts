import { Level, type LogDescriptor } from '../logger.js';

export function gcpErrorProcessor(log: LogDescriptor): LogDescriptor {
  // dont augment anything less than an actual error
  if (log.level < Level.Error) {
    return log;
  }

  // not errorish
  if (!log.data || !('err' in log.data) || !log.data.err) {
    return log;
  }

  const { err, ...data } = log.data;
  const { stack, message, ...restErr } =
    err instanceof Error ? err : Object(err);

  return {
    ...log,
    data: {
      ...data,
      '@type':
        'type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent',
      stack_trace: stack,
      message,
      ...restErr,
    },
  };
}
