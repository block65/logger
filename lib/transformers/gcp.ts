import type { JsonPrimitive } from 'type-fest';
import { Level, LogData, LogDescriptor, Transformer } from '../logger.js';
import { gcpErrorProcessor } from '../processors/gcp.js';
import { safeStringify, withNullProto } from '../utils.js';

type GcpSeverities = 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL';

type GcpJsonLogFormat = {
  severity: GcpSeverities;
  time: string;
  message?: JsonPrimitive;
  ctx?: LogData;
} & LogData;

const logLevelGcpSeverityMap = new Map<Level, GcpSeverities>([
  [Level.Trace, 'DEBUG'],
  [Level.Debug, 'DEBUG'],
  [Level.Info, 'INFO'],
  [Level.Warn, 'WARNING'],
  [Level.Error, 'ERROR'],
  [Level.Fatal, 'CRITICAL'],
]);

export const gcpTransformer: Transformer = (log: LogDescriptor): string => {
  const { level, time, ctx, data, msg } = gcpErrorProcessor(log);

  // See https://cloud.google.com/error-reporting/docs/formatting-error-messages
  const jsonLog: GcpJsonLogFormat = withNullProto({
    severity: logLevelGcpSeverityMap.get(level) || 'INFO',
    time: time.toISOString(),
    message: msg,
    ...data,
    ctx,
  });

  return `${safeStringify(jsonLog)}\n`;
};
