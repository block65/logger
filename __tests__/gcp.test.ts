import {
  afterEach,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { gcpErrorProcessor } from '../lib/processors/gcp.js';
import { createLoggerWithWaitableMock } from './helpers.js';

describe('GCP Processor', () => {
  const initialEnv = process.env;
  beforeEach(() => {
    // jest.clearAllMocks();
    jest.useFakeTimers({ now: new Date('2009-02-13T23:31:30.000Z') });
  });

  afterEach(() => {
    process.env = { ...initialEnv };
    jest.useRealTimers();
  });

  test('Cloud Run', async () => {
    const [logger, callback, errback] = createLoggerWithWaitableMock({
      processors: [gcpErrorProcessor],
    });

    logger.warn(new Error('hello'));
    await logger.end();
    await expect(callback.mock.calls).toMatchSnapshot();
    expect(errback).not.toBeCalled();
  });

  test('Cloud Run variant', async () => {
    process.env.VERSION_NAME = 'logger@foodfacecafe';
    const [logger, callback, errback] = createLoggerWithWaitableMock({
      processors: [gcpErrorProcessor],
    });

    logger.warn(new Error('hello'));
    await logger.end();
    await expect(callback.mock.calls).toMatchSnapshot();
    expect(errback).not.toBeCalled();
  });

  test('Cloud Run Error Object', async () => {
    const [logger, callback, errback] = createLoggerWithWaitableMock({
      processors: [gcpErrorProcessor],
    });

    logger.error(new Error('Ded 1'));
    await logger.end();

    expect(callback.mock.calls).toMatchSnapshot();
    expect(errback).not.toBeCalled();
  });

  test('Cloud Run Fatal with Error Object', async () => {
    const [logger, callback, errback] = createLoggerWithWaitableMock({
      processors: [gcpErrorProcessor],
    });

    logger.fatal(new Error('Ded 2'));
    await logger.end();

    expect(callback.mock.calls).toMatchSnapshot();
    expect(errback).not.toBeCalled();
  });
});
