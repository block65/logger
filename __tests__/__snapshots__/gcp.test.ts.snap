// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GCP Processor Cloud Run 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "hello",
          "name": "Error",
          "stack": "Error: hello
    at Object.<anonymous> (~/logger/__tests__/gcp.test.ts:29:17)
    at Promise.then.completed (~/logger/node_modules/.pnpm/jest-circus@29.5.0/node_modules/jest-circus/build/utils.js:293:28)
    at new Promise (<anonymous>)",
        },
      },
      "err": [Error: hello],
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`GCP Processor Cloud Run Error Object 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "@type": "type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent",
        "message": "Ded 1",
        "name": "Error",
        "stack_trace": "Error: Ded 1
    at Object.<anonymous> (~/logger/__tests__/gcp.test.ts:52:18)
    at Promise.then.completed (~/logger/node_modules/.pnpm/jest-circus@29.5.0/node_modules/jest-circus/build/utils.js:293:28)
    at new Promise (<anonymous>)",
      },
      "err": [Error: Ded 1],
      "level": 50,
      "msg": "Ded 1",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`GCP Processor Cloud Run Fatal with Error Object 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "@type": "type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent",
        "message": "Ded 2",
        "name": "Error",
        "stack_trace": "Error: Ded 2
    at Object.<anonymous> (~/logger/__tests__/gcp.test.ts:64:18)
    at Promise.then.completed (~/logger/node_modules/.pnpm/jest-circus@29.5.0/node_modules/jest-circus/build/utils.js:293:28)
    at new Promise (<anonymous>)",
      },
      "err": [Error: Ded 2],
      "level": 60,
      "msg": "Ded 2",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`GCP Processor Cloud Run variant 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "hello",
          "name": "Error",
          "stack": "Error: hello
    at Object.<anonymous> (~/logger/__tests__/gcp.test.ts:41:17)
    at Promise.then.completed (~/logger/node_modules/.pnpm/jest-circus@29.5.0/node_modules/jest-circus/build/utils.js:293:28)
    at new Promise (<anonymous>)",
        },
      },
      "err": [Error: hello],
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;
